mlir/lib/Quantum/Transforms/annotate_function.cpp

```diff
@@ -0,0 +1,171 @@
+// limitations under the License.
+
+#include "Quantum/IR/QuantumOps.h"
+#include "mlir/Analysis/CallGraph.h"
+#include "mlir/Conversion/LLVMCommon/TypeConverter.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+#include "Catalyst/IR/CatalystOps.h"
+#include "Quantum/Transforms/Passes.h"
+#include "Quantum/Transforms/Patterns.h"
+
+using namespace mlir;
+using namespace catalyst::quantum;
+
+namespace {
+
+static constexpr const char *hasMeasureAttrName = "catalyst.invalidGradientOperation";
+
+bool isAnnotated(func::FuncOp op, const char *attr)
+{
+    return (bool)(op->getAttrOfType<UnitAttr>(attr));
+}
+
+bool invalidGradientOperation(func::FuncOp op)
+{
+    auto res = op.walk([](Operation *o) {
+        if (dyn_cast<MeasureOp>(o) || dyn_cast<catalyst::PythonCallOp>(o) ||
+            dyn_cast<catalyst::CustomCallOp>(o)) {
+            return WalkResult::interrupt();
+        }
+        else {
+            return WalkResult::advance();
+        }
+    });
+    return res.wasInterrupted();
+}
+
+bool successfulMatchLeaf(func::FuncOp op)
+{
+    return !isAnnotated(op, hasMeasureAttrName) && invalidGradientOperation(op);
+}
+
+void annotate(func::FuncOp op, PatternRewriter &rewriter, const char *attr)
+{
+    op->setAttr(attr, rewriter.getUnitAttr());
+}
+
+struct AnnotateFunctionTransform : public OpRewritePattern<func::FuncOp> {
+    using OpRewritePattern<func::FuncOp>::OpRewritePattern;
+
+    LogicalResult match(func::FuncOp op) const override;
+    void rewrite(func::FuncOp op, PatternRewriter &rewriter) const override;
+};
+
+LogicalResult AnnotateFunctionTransform::match(func::FuncOp op) const
+{
+    return successfulMatchLeaf(op) ? success() : failure();
+}
+
+void AnnotateFunctionTransform::rewrite(func::FuncOp op, PatternRewriter &rewriter) const
+{
+    annotate(op, rewriter, hasMeasureAttrName);
+}
+
+std::optional<func::FuncOp> getFuncOp(const CallGraphNode *node, CallGraph &cg)
+{
+    std::optional<func::FuncOp> funcOp = std::nullopt;
+    if (node == cg.getExternalCallerNode())
+        return funcOp;
+    if (node == cg.getUnknownCalleeNode())
+        return funcOp;
+    auto *callableRegion = node->getCallableRegion();
+    funcOp = cast<func::FuncOp>(callableRegion->getParentOp());
+    return funcOp;
+}
+
+std::optional<func::FuncOp> getCallee(CallGraphNode::Edge edge, CallGraph &cg)
+{
+    CallGraphNode *callee = edge.getTarget();
+    return getFuncOp(callee, cg);
+}
+
+bool anyCalleeIsAnnotated(func::FuncOp op, const char *attr, CallGraph &cg)
+{
+    Region &region = op.getRegion();
+    CallGraphNode *node = cg.lookupNode(&region);
+    // TODO: ICE if we do not find the node.
+    for (auto i = node->begin(), e = node->end(); i != e; ++i) {
+        std::optional<func::FuncOp> maybeCallee = getCallee(*i, cg);
+        // An indirect call
+        if (!maybeCallee)
+            return true;
```

Comment:
What is this case and why is it true?

Reply:
For indirect function calls, you don't know the target of the call. This doesn't happen for us. But it is safer to say that if there are indirect function calls which cannot be resolved then they could very likely contain these operations.

Reply:
That makes sense, could you state that in the comment?

Reply:
I added a comment along with a paper for devirtualization methods with different levels of precision. I doubt any of these analyses is already implemented in MLIR, but it is a good summary on the topic. https://github.com/PennyLaneAI/catalyst/pull/204/commits/ab2518c6207692ee27164463da901d4709de2bd8